app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,92 @@
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.UNKNOWN
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+
+    private var isOnPreDrawCalled = false
+    private var startupMode = UNKNOWN
+
+    private var applicationOnCreateTimeStampNanoSeconds: Long? = null
+    private var homeActivityInitTimeStampNanoSeconds: Long? = null
+    private var homeActivityOnRestartTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var homeActivityOnPreDrawTimeStampNanoSeconds: Long = INVALID_TIMESTAMP
+
+    fun onApplicationOnCreate() {
+        startupMode = COLD
+    }
+
+    fun onHomeActivityOnCreate(activityInitNanos: Long) {
+        if (startupMode == UNKNOWN) {
+            startupMode = WARM
+        }
+        this.homeActivityInitTimeStampNanoSeconds = activityInitNanos
+    }
+
+    fun onHomeActivityOnRestart(activityOnRestartNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        startupMode = HOT
+        homeActivityOnRestartTimeStampNanoSeconds = activityOnRestartNanos
+    }
+
+    fun onActivityOnPreDraw(activityOnPreDrawNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        isOnPreDrawCalled = true
+        homeActivityOnPreDrawTimeStampNanoSeconds = activityOnPreDrawNanos
+    }
+
+    /**
+     * if we have both start and finish time for launch, return the difference otherwise return null.
+     */
+    suspend fun getApplicationLaunchTime(): Long? = withContext(Dispatchers.IO) {
+        when {
+            !isOnPreDrawCalled -> {
+                null
+            }
+            else -> {
+                when (startupMode) {
+                    COLD -> {
+                        applicationOnCreateTimeStampNanoSeconds =
+                            stats.getProcessStartTimeStampNano(Process.myPid())
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            applicationOnCreateTimeStampNanoSeconds!!
+                        )
+                    }
+                    WARM -> {
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            homeActivityInitTimeStampNanoSeconds!!
+                        )
+                    }
+                    HOT -> {
+                        homeActivityOnPreDrawTimeStampNanoSeconds.minus(
+                            homeActivityOnRestartTimeStampNanoSeconds!!
+                        )
+                    }
+                    else -> {
+                        null
+                    }
+                }
+            }
+        }
+    }
+
+    enum class StartupMode { COLD, WARM, HOT, UNKNOWN }
+
+    companion object {
+        private const val INVALID_TIMESTAMP: Long = -1
```

Comment:
This isn't used anywhere – why not remove this and use null for the initial value like all the other values?

(Technically, I think it'd be better if everything was non-nullable and started as -1 because otherwise there's autoboxing but it should be negligble; probably no need to name -1 in that case)

Reply:
switched it to null for now